File: StackEvaluation/StackEvaluation/StackEvaluation.cpp
#include<iostream>
#include<stack>
using namespace std;
int StackEvaluation(char *oper) {
	if (*oper=='\0') {
		return 0;
	}
	stack<int> numbers;
	//char *oper[i]=oper;
	for (int i = 0; oper[i];i++) {
		if (isdigit(oper[i]))
		{
			numbers.push(oper[i]-'0');
		}
		else
		{
			int num1 = numbers.top();
			numbers.pop();
			int num2 = numbers.top();
			numbers.pop();
			switch (oper[i])
			{
			case '+':numbers.push(num1 + num2);
				break;
			case '-':numbers.push(num2 - num1);
				break;
			case '*':numbers.push(num1*num2);
				break;
			case '/':numbers.push(num2 / num1);
			}
		}
	}
	return numbers.top();
}
int main() {
	char temp[]="";
	//char temp[] = "231*+9-";
	/*stack<int> keep;
	while (cin>>temp) {
		if (isdigit(*temp)) {
			keep.push(*temp);
		}
		else
		{
			int res=StackEvaluation(keep,temp);
			if (keep.empty()) {
				cout << res << endl;
			}
		}
		
	}*/
	cin >> temp;
	cout<<StackEvaluation(temp)<<endl;
	system("pause");
	return 0;
}

File: py_StackEvaluation/StackEvaluation.py
#用链表实现栈结构
class Node(object):
    def __init__(self,data,next=None):
        self.next=next
        self.data=data
class stack(object):
    def __init__(self,top=None):
        self.top=top
    def push(self,data):
        self.top=Node(data,self.top)
    def pop(self):
        if self.top==None:
            return None
        self.top=self.top.next
    def peek(self):
        return self.top.data if self.top is not None else None
    def isEmpty(self):
        return self.top is None
#使用栈实现后缀计算
def StackEvaluation(inputString):
    if inputString is None:
        print("请输入有效的字串")
        return None
    numbers=stack()
    for ch in inputString:
        if ch.isdigit():
            numbers.push(ch)
        else:
            num1=numbers.peek()
            numbers.pop()
            num2=numbers.peek()
            numbers.pop()
            numbers.push(str(eval(num2+ch+num1)))
    return int(numbers.peek())

string=input("请输入后缀表达式：")
print(StackEvaluation(string))

